bridge-vue: feature (#1)
humphd/web422: Code Examples/week4/bridge-vue/README.md

that you have a better understanding of how to build component based apps, you can
start exploring using the many third-party Vue components that exist.  Whether you
write it, or someone else does, the ideas are the same.

## Search Bar 

> It is an optional augmentation, granted we mind minimal use case for the application.

Reviewer: Let's use some more informal language here, how about:

```suggestion
If you'd like to try a bit more, here's another feature we could add: a search input to filter our bridges.  Thanks to Seneca student Timofei Shchepkin (@fosteman) for adding this!
```

![Screencast of SearchBar in Action](screenshots/search-bar.gif)

> [Semantic-UI-Vue](https://github.com/Semantic-UI-Vue/Semantic-UI-Vue) is a [Semantic-UI](https://semantic-ui.com/) integration, while Semantic is a flexbox friendly UI framework designed for theming. It will bring about pretty elements for input field, loader and error-modal for `bridge-vue` application.

Reviewer: ```suggestion
In addition to extending our app, we'll also gain some experience working with third-party libraries.  Specifically, we'll use [Semantic-UI-Vue](https://github.com/Semantic-UI-Vue/Semantic-UI-Vue). It will bring about pretty elements for the input field, loader and error-modal.
```

### Implementation

> It is as simple as:

Reviewer: ```suggestion
Our change will involve the following steps:
```

- install dependencies
- include the Semantic JS, CSS files

> - apply Semantic middleware to Vue engine

Reviewer: ```suggestion
- apply Semantic middleware to Vue
```

> - rework BridgeMenu component

Reviewer: ```suggestion
- rework the `BridgeMenu` component
```

To install dependencies:
```
npm install semantic-ui-vue semantic-ui-css --save
```

> Then, to make Vue understand Semantic's custom classes and elements, stylesheet and JS module files are imported into `main.js`:

Reviewer: ```suggestion
Then, to make Vue understand Semantic's custom classes and elements, we'll import the necessary stylesheet and JS module files in `main.js`:
```

> ```

Reviewer: Use ```js here to get syntax highlighting.

import SuiVue from 'semantic-ui-vue';
Vue.use(SuiVue);
import 'semantic-ui-css/semantic.min.css';
```

> <strong>Template markup</strong> of `BridgeMenu.vue` component is modified accordingly. 

Reviewer: This can be turned into a heading, maybe ### Updates to `BridgeMenu.vue`

> First, outer container `<div class='main'>` is now <em>classified</em>, instead of <em>identified</em>,  as `menu`, for the reason of CSS rules to be reused. 

Reviewer: ```suggestion
First, our outer container `<div class='main'>` now uses the `menu` CSS class instead of an `id`.
```

In this markup, conditional directives `v-if`, `v-else` replace `v-bind`ing of class-attributes (i.e. `active`) that Semantic uses for the sake of readability. 

> Lastly, when menu-items are being deployed into the listing, the source array is changed to a new stateful property `this.showBridges`, that is a filtered version of `this.bridges`. 

Reviewer: ```suggestion
Lastly, in order to generate our `menu-item` components, we switch to using `this.showBridges`, which is a filtered version of `this.bridges`. 
```

```angular2html
<template>
    <div class="menu">
        <!--Search Field-->

>         <div style="width:100%;" class="ui icon input">

Reviewer: What do you think about moving this inline style to the `<styles>` block and adding a class instead, to keep the approach consistent here?

Author: Grace to `<style scoped>`, `width` can be safely assigned to `input` class

            <input placeholder="Name a bridge..." v-model="search" />
            <i class="search icon"></i>
        </div>
        <div class="menu">
            <!--Loader-->
            <div v-if="status.loading" class="menu ui segment inverted">
                <div class="ui active dimmer">
                    <div class="ui text loader">Loading bridges...</div>
                </div>
            </div>
            <!--Error Modal-->
            <div v-if="status.errored">
                <div class="ui warning message">
                    <div class="header">
                        Error: Data was not loaded.
                    </div>
                    Hint: Check API endpoint!
                </div>
            </div>
            <!--BridgeListing-->
            <div v-else>
                <menu-item

>                         v-for="bridge of shownBridges" <!--filtered array-->

Reviewer: Question: why `shownBridges` and not `filteredBridges`?  That they are shown is more about what we *do* with that data vs. what it is.

Author: @humphd, your suggestion is atmost relevant. Even on the highest level, we're obviously tackling declarative Vue.js

                        :key="bridge.id"
                        :bridge="bridge"
                        @click="bridgeSelected"
                />
            </div>
        </div>
    </div>
</template>
```

Small change to `<style>`, to suit multiple 'menu'-like elements:
```angular2html
<style scoped>
    .menu {
        height: 100%;
    }
</style>
```

> En fin, under `<script>`, we need to declare 2 new variables to the state: `search` and `shownBridges`. Former will be synced (`v-model`) with SearchBar, while latter will be dynamically updated with filtered array of bridges. 

Reviewer: ```suggestion
In our `<script>` block, we need to declare two new `data` variables: `search` and `shownBridges`. The Former will be synced (`v-model`) with `SearchBar`, while latter will be dynamically updated with the filtered array of bridges.
```

> To filter out `shownBridges`, `search` property will be <em>watched</em>, so that, whence new value is entered, the listing is rendered. <em>Notably</em>, original array stays in place, even more so, it is going to be automatically updated in case of fresh data downloaded from backend.

Reviewer: First, a question: why not just use a `computed` value here for the filtered bridges, which depends on both `this.search` and `this.bridges`?  It would return the filtered array of bridges when either/both updates.  I think that's more idiomatic of Vue.

Author: @humphd, computed values permit no mutation from inside. `filteredBridges` will remain in state, however,` v-for of _computed_` is malformed.
```jsx harmony
computed: {
filter: function() {
      return this.bridges.filter(bridge =>
         bridge.name.toLowerCase()
           .includes(this.search.toLowerCase()));
     }
}
...
<div v-else>
        <menu-item
               v-for="(bridge, index) of **???**"
                :key="filter[index].bridge.id"
                :bridge="filter[index].bridge"
                 @click="bridgeSelected"
                />
 </div>
```
I foresee use of `menu-item` interface with `filter` method in it, but how would you implement this?

Reviewer: ` v-for="(bridge, index) of filteredBridges"` and rename your `computed` property to `filteredBridges`

                this.shownBridges = this.bridges.filter(bridge =>
                bridge.name.toLowerCase()
                    .includes(this.search.toLowerCase()));
            }
        },
        components: {
            MenuItem
        },
        created: function() {
            this.loadBridges();
        },
        methods: {
            loadBridges: function() {
                this.status.loading = true;

                // Use our bridge.js function to talk to the REST API.
                getBridgeData()
                    .then(bridges => {
                        this.status.loading = false;
                        this.bridges = bridges;
                    })
                    .catch(err => {
                        console.error('Unable to load bridge data', err.message);
                        this.status.errored = true;
                    });
            },
            bridgeSelected: function(bridge) {
                // When the user clicks a menu item, emit a `change`
                // event for the menu control, along with bridge value
                this.$emit('change', bridge);
            }
        }
    }
</script>
```

> ### Augmentation Conclusion

Reviewer: We can get rid of this header and just have the final sentence.
